app/lib/goods_nomenclature_mapper.rb

```diff
@@ -52,20 +52,19 @@ def for_goods_nomenclature(ref_goods_nomenclature)
   private
 
   def process
-    # first pair
-    traverse(goods_nomenclatures, nil, goods_nomenclatures.first)
-    # all other pairs
-    traverse(goods_nomenclatures, goods_nomenclatures.first, goods_nomenclatures.second)
+    traverse(goods_nomenclatures, goods_nomenclatures.first)
+    traverse(goods_nomenclatures, goods_nomenclatures.second)
   end
 
-  def traverse(goods_nomenclatures, primary, secondary)
-    # ignore case when first goods_nomenclature is blank it's a direct child of the heading
-    unless goods_nomenclatures.index(secondary).blank?
-      next_goods_nomenclature = goods_nomenclatures[goods_nomenclatures.index(secondary) + 1]
-      if next_goods_nomenclature.present? # we are not at the end of the goods_nomenclature array
-        map_goods_nomenclatures(secondary, next_goods_nomenclature)
-        traverse(goods_nomenclatures, secondary, next_goods_nomenclature)
-      end
+  def traverse(goods_nomenclatures, current_goods_nomenclature)
+    current_good = current_goods_nomenclature
+
+    while (index_gn = goods_nomenclatures.index(current_good))
```

Comment:
This is basically just an `each_cons(2)` no? I've basically never seen a need for while in ruby unless the condition for exiting the loop is outside of the data-structure.

Reply:
yeah, I thought about it, in this first step I strictly followed the logic in the recursion, but it seems to be just a each loop, I'll change it.